File: 20. Valid Parentheses/valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        dict = {'(':')', '[':']', '{':'}'}
        stack = []
        while(s):
            if s[0] in dict:
                stack.append(dict[s[0]])
                s = s[1:]
            else:
                if stack and s[0] == stack[-1]:
                    s = s[1:]
                    stack.pop(-1)
                else:
                    return False
        if not stack:
            return True
        else:
            return False

File: 20. Valid Parentheses/valid-parentheses.c
char getHalfParentheses(char c){
    switch(c){
        case ')':
            return '(';
        case ']':
            return '[';
        case '}':
            return '{';
        default:
            return '0';
    }
}
bool isValid(char * s){
    int len_s = strlen(s), i, index_stack=0;
    char *stack = malloc(sizeof(char)*len_s);
    for(i=0; i<len_s; i++){
        if(s[i] == '(' || s[i] == '[' || s[i] == '{')
            stack[index_stack++] = s[i];
        else{
            if(index_stack > 0 && stack[index_stack-1] == getHalfParentheses(s[i]))
                index_stack--;
            else
                return false;
        }
    }
    return index_stack == 0 ? true : false;
}
